[WIP] whitelist

Igmat opened the pull request:
@jridgewell, I've created this PR to have an ability to comment your fork, since I have a few questions to it.

1. Is `Proxy with whitelist` discussed anywhere as an approach?
2. Is such `Proxy` tunnels `Symbol.private` by default, but if this `Symbol` is whitelisted it traps it?

jridgewell committed:
Update to whitelist

jridgewell committed:
Add sanity checks

Igmat commented on index.js:
```diff
@@ -130,21 +112,17 @@ function createWrapFn(originalsToProxies, proxiesToOriginals, unwrapFn) {
-
-    const wrap = createWrapFn(originalProxies, originalTargets, unwrap);
-    const wrapOuter = createWrapFn(outerProxies, originalProxies, wrap)
+function membrane(left, right) {
```
Such approach seems to be not an original `Membrane`, since in time of creation we have to be aware of both membrane sides.
I'm not sure that such changed pattern covers all the original use-cases. Am I mistaken?

jridgewell replied:
> Such approach seems to be not an original Membrane, since in time of creation we have to be aware of both membrane sides.

I find it easier to reason about membranes this way. You need not expose a "right" object graph (it can be a blank object), or even the `wrappedRightSide`. But, this simplifies the setup of the Right object graph. Without this, you would need:

```js
const leftPriv = Symbol('left private'); // Symbol.private();
class Left {
    // [leftPriv] = some value
};
const leftValue = Left.leftValue = {
    leftValue: true,
};
Left.leftPriv = leftPriv;

const rightPriv = Symbol('right private');
class Right {
    // [rightPriv] = some value
}
const rightValue = Right.rightValue = {
    rightValue: true,
};
Right.rightPriv = rightPriv;


const graph = {
  Left,
};
const wrappedGraph = membrane(graph);
wrappedGraph.Right = Right;

assert(graph.Left is unwrapped);
assert(graph.Right is wrapped proxy);
assert(wrappedGraph.Left is wrapped proxy);
assert(graph.Right is unwrapped);
```

> I'm not sure that such changed pattern covers all the original use-cases. Am I mistaken?

They both reduce to the same thing. (see above)

But you need to be careful when preforming the assertions. When you read the invariants, they're split into both a set then an get assertion, but also by "sides".

The left side of the slide has access to `Left` and `wrappedRightSide`, but not `Right` or `wrappedLeftSide`. The right side of the slide has access to `Right` and `wrappedLeftSide`, but not `Left` or `wrappedRightSide`. You can see this by going to the next slide, the one with the diagram.

hax commented:
"Proxy with whitelist" has been discussed several years ago, but don't know why finally been dropped.

hax commented:
This whitelist implementation has a hole that the whitelist can be proxied, so the private symbols can be leaked when `whitelist.has(privSymbol)` is called internally.

The API should be reversed, not accept a whitelist, but expose a whitelist which is a wrapped Set.

jridgewell reviewed (commented):
> Is Proxy with whitelist discussed anywhere as an approach?

I found it while researching in esdiscuss. I think @rdking has also brought it up in the few issue threads (and I'm guessing also found it in esdiscuss?). I see it only talked about in https://github.com/tc39/tc39-notes/blob/master/es6/2012-09/sept-19.md#proxy-and-private-names, and then never mentioned again. It's on the old [doku wiki](https://web.archive.org/web/20160317165828/http://wiki.ecmascript.org/doku.php?id=strawman:proxies_names).

> Is such Proxy tunnels Symbol.private by default, but if this Symbol is whitelisted it traps it?

Yup.

jridgewell committed:
Membrane with source graph tracking

jridgewell committed:
Whitelist support

jridgewell committed:
Refactor test suites

jridgewell committed:
Implement whitelist proxies

jridgewell committed:
Fix bug

jridgewell committed:
Normalize line endings

jridgewell committed:
Remove unnecessary params

jridgewell committed:
Add shadow target and regular target tests

jridgewell committed:
Use same object graph in handlePrivates

The private data is guaranteed to be from the same object graph as the proxy's target.

jridgewell committed:
Comments

jridgewell committed:
Rename a few variables

jridgewell committed:
Fix membrane to recursively wrap as new privates are found

A private symbol could have been stored as the value of a different
private symbol'd property.

jridgewell committed:
Complete the membrane implementation

jridgewell committed:
Add prototype chain tests

jridgewell committed:
Add priveleged ShadowTarget API

jridgewell committed:
Cleanup tests

jridgewell committed:
Cleanup test assertions

jridgewell committed:
ProtoProxy exotic object

jridgewell committed:
Test all possible configurations

jridgewell committed:
Sync the prototype chain

If the prototype chain is in sync, no additional proxy wrappers are
necessary.
